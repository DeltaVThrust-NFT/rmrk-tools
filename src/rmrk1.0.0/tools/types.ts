import { Send } from "../classes/send";
import { Buy } from "../classes/buy";
import { List } from "../classes/list";
import { NFT } from "../classes/nft";
import { Emote } from "../classes/emote";
import { ChangeIssuer } from "../classes/changeissuer";

export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
  remark: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
  extras?: BlockCall[];
};

export enum OP_TYPES {
  BUY = "BUY",
  LIST = "LIST",
  MINT = "MINT",
  MINTNFT = "MINTNFT",
  SEND = "SEND",
  EMOTE = "EMOTE",
  CONSUME = "CONSUME",
  CHANGEISSUER = "CHANGEISSUER",
}

export type Interaction = Send | Buy | List | Emote | ChangeIssuer;
