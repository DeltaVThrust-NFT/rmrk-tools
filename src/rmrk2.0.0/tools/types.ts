import { Send } from "../classes/send";
import { Buy } from "../classes/buy";
import { List } from "../classes/list";
import { Emote } from "../classes/emote";
import { ChangeIssuer } from "../classes/changeissuer";

export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
  remark: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
  extras?: BlockCall[];
};

export type BaseType = "svg";

export type Interaction = Send | Buy | List | Emote | ChangeIssuer;
