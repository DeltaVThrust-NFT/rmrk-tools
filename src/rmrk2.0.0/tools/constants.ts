export const VERSION = "2.0.0";
export const PREFIX = "RMRK";

export enum OP_TYPES {
  BUY = "BUY",
  LIST = "LIST",
  CREATE = "CREATE",
  MINT = "MINT",
  SEND = "SEND",
  EMOTE = "EMOTE",
  CHANGEISSUER = "CHANGEISSUER",
  CONSUME = "CONSUME",
  BASE = "BASE",
  EQUIPPABLE = "EQUIPPABLE",
  RESADD = "RESADD",
  ACCEPT = "ACCEPT",
  EQUIP = "EQUIP",
}
