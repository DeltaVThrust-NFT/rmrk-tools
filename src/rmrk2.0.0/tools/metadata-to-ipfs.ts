import fs from "fs";
// @ts-ignore
import pinataSDK from "@pinata/sdk";
import { NFTMetadata } from "../classes/nft";

const pinata = pinataSDK(process.env.PINATA_KEY, process.env.PINATA_SECRET);

const defaultOptions = {
  pinataOptions: {
    cidVersion: 1,
  },
};

export const pinToIpfs = async (filePath: string, name?: string) => {
  const options = { ...defaultOptions, pinataMetadata: { name } };
  try {
    const readableStreamForFile = fs.createReadStream(filePath);
    const result = await pinata.pinFileToIPFS(readableStreamForFile, options);
    return result.IpfsHash;
  } catch (error: any) {
    console.error(error);
  }
};

export const uploadRMRKMetadata = async (
  imagePath: string,
  metadataFields: NFTMetadata
): Promise<string> => {
  const options = {
    ...defaultOptions,
    pinataMetadata: { name: metadataFields.name },
  };
  try {
    const imageHash = await pinToIpfs(imagePath, metadataFields.name);
    const metadata = { ...metadataFields, image: `ipfs://ipfs/${imageHash}` };
    const metadataHashResult = await pinata.pinJSONToIPFS(metadata, options);
    return `ipfs://ipfs/${metadataHashResult.IpfsHash}`;
  } catch (error: any) {
    return "";
  }
};
