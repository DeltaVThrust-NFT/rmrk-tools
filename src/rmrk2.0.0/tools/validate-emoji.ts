import stringFromCodePoint from "./polyfill-string-from-codepoint";
import emojiRegex from "emoji-regex/text";

/**
 * Converted dashed emoji unicodes into native emoji unicode with prefix
 * @param unified - emoji unicode
 */
export const unifiedToNative = (unified: string) => {
  try {
    const unicodes = unified
      .replace(/([U+]){2}/g, "")
      .split(unified.includes("-") ? "-" : " ");
    const codePoints = unicodes.map((u) => (u ? `0x${u}` : ""));

    return stringFromCodePoint(
      ...codePoints.map((codePoint) => Number(codePoint))
    );
  } catch (error: any) {
    return "";
  }
};

/**
 * Validate emoji
 * @param emoji
 */
export const isValidEmoji = (emoji: string) => {
  const unified = unifiedToNative(emoji);
  const regex = emojiRegex();
  return regex.test(unified);
};
