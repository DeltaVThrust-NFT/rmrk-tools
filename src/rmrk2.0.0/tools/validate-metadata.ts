import { NFTMetadata } from "../classes/nft";
import { CollectionMetadata } from "../classes/collection";
import {
  optional,
  pattern,
  string,
  type,
  any,
  assert,
  object,
  enums,
  record,
  literal,
} from "superstruct";
import { IProperties } from "./types";

const MetadataStruct = type({
  name: optional(string()),
  description: optional(string()),
  image: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
  animation_url: optional(
    pattern(string(), new RegExp("^(https?|ipfs)://.*$"))
  ),
  image_data: optional(string()),
  background_color: optional(string()),
  youtube_url: optional(pattern(string(), new RegExp("^https://.*$"))),
  attributes: any(),
  external_url: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
});

export const PropertiesStruct = object({
  value: any(),
  type: enums(["string", "array", "object", "int", "float"]),
  _mutable: optional(literal("freeze")),
});

export const validateAttributes = (properties?: IProperties) => {
  if (!properties) {
    return true;
  }
  assert(properties, record(string(), PropertiesStruct));

  Object.values(properties).forEach((attribute) => {
    const { value, type, _mutator } = attribute;
    if (type === "string") {
      if (typeof value !== "string") {
        throw new Error("for type 'string' 'value' has to be a string");
      }
    }

    if (type === "int" || type === "float") {
      if (typeof value !== "number") {
        throw new Error("for type 'number' 'value' has to be a number");
      }
    }

    if (type === "array") {
      if (!Array.isArray(value)) {
        throw new Error("for type 'array' 'value' has to be an array");
      }
    }

    if (type === "object") {
      if (typeof value !== "object") {
        throw new Error("for type 'object' 'value' has to be an Object");
      }
    }
  });

  return true;
};

/**
 * Validate Metadata according to OpenSea docs
 * https://docs.opensea.io/docs/metadata-standards
 * @param metadata
 */
export const validateMetadata = (
  metadata: NFTMetadata | CollectionMetadata
) => {
  assert(metadata, MetadataStruct);

  if (!metadata.image && !(metadata as NFTMetadata).animation_url) {
    throw new Error("image or animation_url is missing");
  }

  validateAttributes(metadata.properties);
  return true;
};
