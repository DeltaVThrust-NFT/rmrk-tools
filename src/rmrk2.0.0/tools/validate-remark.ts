import { OP_TYPES, PREFIX, VERSION } from "./constants";
import {
  assert,
  define,
  string,
  pattern,
  number,
  optional,
  type,
  is,
  enums,
  union,
  array,
} from "superstruct";
import { getRemarkData } from "./utils";
import { collectionRegexPattern } from "../classes/equippable";

const PartStruct = type({
  type: enums(["slot", "fixed"]),
  unequip: optional(enums(["unequip", "burn"])),
  z: optional(number()),
  src: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
  id: string(),
  equippable: optional(union([string(), array(string())])),
});

const CollectionStruct = type({
  max: number(),
  issuer: string(),
  symbol: string(),
  id: string(),
  metadata: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
});

const NFTStruct = type({
  collection: string(),
  symbol: string(),
  transferable: number(),
  sn: string(),
  metadata: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
});

const ResourceStruct = type({
  base: optional(pattern(string(), new RegExp("^base-"))),
  src: optional(string()),
  slot: optional(pattern(string(), new RegExp(/^base-\S+\.{1}\S+$/))),
  metadata: optional(pattern(string(), new RegExp("^(https?|ipfs)://.*$"))),
});

const IsBigInt = define("BigInt", (value: any) => {
  try {
    if (!is(value, string())) {
      return false;
    }

    const priceBigInt = BigInt(parseInt(value));
    return typeof priceBigInt === "bigint";
  } catch (error) {
    return false;
  }
});

const LISTStruct = type({
  id: string(),
  price: IsBigInt,
});

const ACCEPTStruct = type({
  id: string(),
  nftId: string(),
  entity: enums(["NFT", "RES"]),
});

const EQUIPStruct = type({
  id: string(),
  baseslot: union([pattern(string(), new RegExp(/^$/)), pattern(string(), new RegExp(/^base-\S+\.{1}\S+$/))]), // Allow only 1 dot
});

const BUYStruct = type({
  id: string(),
});

const CONSUMEStruct = type({
  id: string(),
});

const SENDStruct = type({
  id: string(),
  recipient: string(),
});

const EMOTEStruct = type({
  id: string(),
  unicode: string(),
  namespace: enums(["RMRK1", "RMRK2", "PUBKEY"]),
});

const BaseStruct = type({
  issuer: string(),
  type: enums(["svg"]),
  symbol: string(),
  parts: array(PartStruct),
});

const CHANGEISSUERStruct = type({
  id: string(),
  issuer: string(),
});

const EQUIPPABLEStruct = type({
  id: string(),
  slot: string(),
  classes: pattern(string(), new RegExp(collectionRegexPattern)),
});

export const validateRemarkBase = (remark: string, opType: OP_TYPES) => {
  const [prefix, op_type, version] = remark.split("::");
  if (prefix.toUpperCase() !== PREFIX) {
    throw new Error("Invalid remark - does not start with RMRK");
  }
  if (op_type !== opType) {
    throw new Error(`The op code needs to be ${opType}, but it is ${op_type}`);
  }
  if (version !== VERSION) {
    throw new Error(
      `This remark was issued under version ${version} instead of ${VERSION}`
    );
  }
};

export const validateCollection = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, dataString] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.CREATE);
    const obj = getRemarkData(dataString);
    if (!obj.metadata) {
      throw new Error("NFT Collection is missing metadata");
    }
    return assert(obj, CollectionStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateNFT = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, dataString] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.MINT);
    const obj = getRemarkData(dataString);
    return assert(obj, NFTStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateBase = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, dataString] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.BASE);
    const obj = getRemarkData(dataString);
    return assert(obj, BaseStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during Base remark validation"
    );
  }
};

export const validateList = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, price] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.LIST);
    return assert({ id, price }, LISTStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateResadd = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, resource] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.RESADD);
    if (!id) {
      throw new Error("No NFT id specified for RESADD");
    }
    const obj = getRemarkData(resource);
    return assert(obj, ResourceStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during RESADD remark validation"
    );
  }
};

export const validateSend = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, recipient] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.SEND);
    if (/\s/g.test(recipient)) {
      throw new Error(
        "Invalid remark - No whitespaces are allowed in recipient"
      );
    }
    return assert({ id, recipient }, SENDStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateEmote = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, unicode] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.EMOTE);
    return assert({ id, unicode }, EMOTEStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateEquippable = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, slot, classes] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.EQUIPPABLE);
    return assert({ id, slot, classes }, EQUIPPABLEStruct);
  } catch (error) {
    throw new Error(
      error?.message ||
        "Something went wrong during EQUIPPABLE remark validation"
    );
  }
};

export const validateChangeIssuer = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, issuer] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.CHANGEISSUER);
    return assert({ id, issuer }, CHANGEISSUERStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateBuy = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.BUY);
    return assert({ id }, BUYStruct);
  } catch (error) {
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateConsume = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.CONSUME);
    return assert({ id }, CONSUMEStruct);
  } catch (error) {
    console.log("StructError is:", error);
    throw new Error(
      error?.message || "Something went wrong during remark validation"
    );
  }
};

export const validateAccept = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, nftId, entity, id] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.ACCEPT);
    return assert({ id, nftId, entity }, ACCEPTStruct);
  } catch (error) {
    console.log("StructError is:", error);
    throw new Error(
      error?.message || "Something went wrong during ASCCEPT remark validation"
    );
  }
};

export const validateEquip = (remark: string): any => {
  // With array destructuring it's important to not remove unused destructured variables, as order is important
  const [_prefix, _op_type, _version, id, baseslot = ""] = remark.split("::");

  try {
    validateRemarkBase(remark, OP_TYPES.EQUIP);
    return assert({ id, baseslot }, EQUIPStruct);
  } catch (error) {
    console.log("StructError is:", error);
    throw new Error(
      error?.message || "Something went wrong during EQUIP remark validation"
    );
  }
};
