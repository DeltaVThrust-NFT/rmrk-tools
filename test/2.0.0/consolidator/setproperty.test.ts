import { Consolidator, NFT } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  createCollectionMock2,
  getAliceKey,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock2,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";
import { OP_TYPES } from "../../../src/rmrk2.0.0/tools/constants";
import { stringToHex } from "@polkadot/util";

beforeAll(async () => {
  return await cryptoWaitReady();
});

const mintNftWithPropertiesOwner = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutation: {
          allowed: true,
        },
        value: "foo",
        type: "string",
      },
    },
  });

const mintGemNft = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock2().id,
    symbol: "KANGEM",
    sn: "111".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
  });

const mintNftWithMutatorCondition = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "888".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutation: {
          allowed: true,
          with: {
            opType: OP_TYPES.BURN,
            condition: "d43593c715a56da27d-KANARIAGEMS",
          },
        },
        value: "foo",
        type: "string",
      },
    },
  });

const mintNftWithProperties = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        value: "foo",
        type: "string",
      },
      royaltyInfo: {
        type: "royalty",
        value: {
          receiver: "xxx",
          royaltyPercentFloat: 0.2,
        },
        _mutation: {
          allowed: true,
        },
      },
    },
  });

describe("rmrk2.0.0 Consolidator: SETPROPERTY", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
    ...getBlockCallsMock(createCollectionMock2().create()),
  ];

  it("should throw if we are trying to mutate immutable property", async () => {
    expect(() => mintNftWithProperties(5).setproperty("test", 2)).toThrow();
  });

  it("should throw if we are trying to mutate property with wrong rootowner _mutation", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar"),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to set property on a non-owned NFT. Expected 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty but received 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call passed", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to mutate property without matching extra call of op type BURN"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call passed", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintNftWithMutatorCondition().mint()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "The op code needs to be BURN, but it is MINT"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call condition", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintNftWithPropertiesOwner(5).burn()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should allow if we are trying to mutate property with matching mutation condition", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(mintGemNft().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintGemNft(5).burn()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should allow to mutate property if owner _mutator match", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).send(getBobKey().address)
      ), // Send to Bob first
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar"),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should be able to freeze property ", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar", "freeze")
      ),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "baz")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to set property on immutable property test"
    );
  });

  it("should not allow to mutate royalty ", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithProperties().mint()),
      ...getBlockCallsMock(
        mintNftWithProperties(4).setproperty("royaltyInfo", {
          receiver: "xxx",
          royaltyPercentFloat: 0.3,
        })
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to mutate an attribute of type 'royalty'."
    );
  });
});
