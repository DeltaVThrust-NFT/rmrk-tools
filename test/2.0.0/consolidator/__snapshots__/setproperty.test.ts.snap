// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should allow if we are trying to mutate property with matching mutation condition 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
            "with": Object {
              "condition": "d43593c715a56da27d-KANARIAGEMS",
              "opType": "BURN",
            },
          },
          "type": "string",
          "value": "bar",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIAGEMS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIAGEMS-KANGEM-00000111",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000111",
      "symbol": "KANGEM",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should allow to mutate property if owner _mutator match 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 5,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
          },
          "type": "string",
          "value": "bar",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should throw if we are trying to mutate property with no matching extra call condition 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 6,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[SETPROPERTY] Attempting to mutate an property but it's mutation condition doesn't match",
      "object_id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "op_type": "SETPROPERTY",
    },
  ],
  "nfts": Array [
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
            "with": Object {
              "condition": "d43593c715a56da27d-KANARIAGEMS",
              "opType": "BURN",
            },
          },
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
          },
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;
